README.md

```diff
@@ -335,192 +357,227 @@ module with string literals may fail due to lack of memory resources,
+`string.measure_utf8`, if the string contains an isolated surrogate,
+return -1.
+
+The maximum number of bytes returned by these instructions is
+2<sup>31</sup>-1.  If an encoding would require more bytes, it is as if
+the contents can't be encoded at all (the return value is -1).
 
 ```
-(string.end_wtf16 str:stringref)
-  -> pos:i32
+(string.encode_utf8 $memory str:stringref ptr:address)
+(string.encode_wtf8 $memory str:stringref ptr:address)
+(string.encode_wtf16 $memory str:stringref ptr:address)
 ```
-Return the position after the last code unit in the WTF-16 encoding of
-the string *str*.
+Encode the contents of the string *`str`* as UTF-8, WTF-8, or WTF-16,
+respectively, to memory at *`ptr`*.  The number of bytes written will be
+the same as returned by the corresponding `string.measure_*encoding*`.
+Note that no `NUL` terminator is ever written.  For
+`string.encode_utf8`, trap if the string contains an isolated surrogate.
 
-For WTF-8 positions, use the `string.advance_wtf8` and
-`string.rewind_wtf8` seek instructions:
+The maximum number of bytes that can be encoded at once by
+`string.encode` is 2<sup>31</sup>-1.  If an encoding would require more
+bytes, it is as if the codepoints can't be encoded (a trap).
+
+### Concatenation
 
 ```
-(string.advance_wtf8 str:stringref pos:i32 codepoints:i32)
-  -> pos:i32, codepoints:i32
-(string.rewind_wtf8 str:stringref pos:i32 codepoints:i32)
-  -> pos:i32, codepoints:i32
+(string.concat a:stringref b:stringref) -> stringref
 ```
-Return a new WTF-8 position for the string *`str`* which is
-*`codepoints`* codepoints after or before *`pos`*.  Also return the
-number of codepoints by which the new position differs from the old,
-which may be less than the *`codepoints`* parameter if end-of-string or
-beginning-of-string would be reached.  The *`codepoints`* argument is
-interpreted as a `u32`.
+Return a new stringref containing the codepoints from *`a`* followed by
+the codepoints from *`b`*.
 
-#### Invalid position values
+Note that the implementation should take care when, at any future time,
+treating the resulting string as a sequence of codepoints.  If *`a`*'s
+last codepoint is a high surrogate and *`b`*'s first codepoint is a low
+surrogate, these two codepoints combine into one, as if they were the
+two code units of a UTF-16-encoded unicode scalar value.
 
-All instructions which take string positions trap if the position is not
-valid for the given string.
+Concatenating two strings is a form of dynamic allocation and can fail.
+If an allocation fails, the implementation must trap.  Fallible
+`string.concat` is a possible future extension.
 
-#### Position value limits
+### Predicates
 
-The maximum valid `i32` position value is 2<sup>31</sup>.
+```
+(string.eq a:stringref b:stringref) -> i32
+```
+Return 1 if the strings *`a`* and *`b`* contain the same codepoint
+sequence.  Return 0 otherwise.
+
+```
+(string.is_usv_sequence str:stringref)
```

Comment:
Looking at this again.  In a component-model scenario where there is an inter-component function call to an interface taking a string, and that both sides actually implement strings with stringrefs, in that case we can pass the stringref value directly without copying -- but only if it's a USV sequence.  We don't actually need to compute the length in that case.  I would go ahead and include it, as it will make component-model users happy.